bullet-featherstone: Support nested models

iche033 opened the pull request:
# 🎉 New feature

Closes https://github.com/gazebosim/gz-physics/issues/546

## Summary

Extended bullet-featurestone implementation to support these features related to nested models:
* `ConstructSdfJoint` (limited to creating world joints)
* `ConstructSdfNestedModel` (https://github.com/gazebosim/gz-physics/issues/546)
* `GetNestedModelFromModel`
* `WorldModelFeature` (https://github.com/gazebosim/gz-physics/issues/547)
* `RemoveEntities`

The nested models are currently added to a single multibody since bullet-featherstone does not seem to support joints between 2 different multibodies. The nested model just keeps a shared ptr to the rootLink's btMultibody.

## Test it

With the addition of the above features, more common tests are now being run using bullet-feathersone plugin. Note that I had to skip a few of tests in `free_joint_features` and `joint_features` because the bullet-featherstone implementation does not support 1) multiple floating bodies / sub-trees yet (https://github.com/gazebosim/gz-physics/issues/550) and 2) joints between two different models.

Here's an example [nested model world file](https://gist.github.com/iche033/dd30ba98d92fde02a643fa94c5e2f6a9) I use for testing. To run:

```
gz sim -v 4 -r nested.sdf --physics-engine gz-physics-bullet-featherstone-plugin
```


## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

scpeters reviewed (commented):
It's a little weird that the `WorldModelAPI` test in world_features.cc passes even though it's not actually creating the joints between `//world/model`s. When I load `world_joint_test.sdf` in `gz sim` with bullet-featherstone, I see the following error messages, which aren't obvious to interpret but do indicate that something is wrong:

~~~
[Err] [Physics.cc:1710] Attempting to create a new joint [j1], but the physics engine doesn't support constructing joints at runtime.
[Err] [Physics.cc:1710] Attempting to create a new joint [j2], but the physics engine doesn't support constructing joints at runtime.
~~~

I also tried loading the `nested_model.sdf` example world from `gz-sim`, which I believe is not supported because it has multiple kinematic trees, but it results in a comprehensible error message that is obscured by infinite console spamming:

~~~
[Err] [SDFFeatures.cc:534] Multiple subt-trees / floating links detected in a model. This is not supported in bullet-featherstone implementation yet.
[Wrn] [Physics.cc:1285] Cannot create a new link [link_01] because the physics engine plugin does not support link construction during runtime
[Wrn] [Physics.cc:1372] Collision's parent entity [13] not found on link map.
[Err] [Physics.cc:2894] Internal error: link [13] not in entity map
[Err] [Physics.cc:3050] Internal error: entity [13] not in entity map
...
other messages
...
[Err] [Physics.cc:2894] Internal error: link [13] not in entity map
[Err] [Physics.cc:2894] Internal error: link [13] not in entity map
[Err] [Physics.cc:2894] Internal error: link [13] not in entity map
[Err] [Physics.cc:2894] Internal error: link [13] not in entity map
...
~~~

This may be a gz-sim problem but if there's a way to mitigate it in gz-physics, that would be helpful to users.

I also get infinite console spam when loading the nested_model_joint_positions.sdf example world from gz-sim unless I remove `model4`

iche033 commented:
> It's a little weird that the `WorldModelAPI` test in world_features.cc passes even though it's not actually creating the joints between `//world/model`s

yeah the `WorldModelAPI` test checks the parent-child relationship between world and models only. It does not actually require a world joint to be present for a model to be a child of the 'world model'.

> When I load `world_joint_test.sdf` in gz sim with bullet-featherstone, I see the following error messages, which aren't obvious to interpret but do indicate that something is wrong:

I added support for the `ConstructSdfJoint` feature which is used by gz-sim for creating a `//world/joint`. This new function is  currently also only limited to creating world joints (and not other joints between or within models). [2c85f5e](https://github.com/gazebosim/gz-physics/pull/574/commits/2c85f5ea8b8357d16f450fad9b2057bfe486157c)

> I also tried loading the nested_model.sdf example world from gz-sim, which I believe is not supported because it has multiple kinematic trees, but it results in a comprehensible error message that is obscured by infinite console spamming:

I added necessary logic to add all remaining links (floating bodies / other subtrees) in gz-physics but without bullet structures. This is just for book-keeping and prevents gz-sim from printing warning messages when trying to access these links. Added warning msgs to let users know that these links are non-functional. [2c85f5e](https://github.com/gazebosim/gz-physics/pull/574/commits/2c85f5ea8b8357d16f450fad9b2057bfe486157c) and [06016b7](https://github.com/gazebosim/gz-physics/pull/574/commits/06016b79d2837522d7fee8907bc67d28a1d17ea1)

> I also get infinite console spam when loading the nested_model_joint_positions.sdf example world from gz-sim unless I remove model4

Fixed issue with an empty top level model that has no links. [2c85f5e](https://github.com/gazebosim/gz-physics/pull/574/commits/2c85f5ea8b8357d16f450fad9b2057bfe486157c)

iche033 commented on bullet-featherstone/src/SDFFeatures.hh:
```diff
@@ -59,6 +65,13 @@ class SDFFeatures :
       const Identity &_linkID,
       const ::sdf::Collision &_collision) override;
+
+  /// \brief Construct a bullet entity from a sdf::model
```
[2c85f5e](https://github.com/gazebosim/gz-physics/pull/574/commits/2c85f5ea8b8357d16f450fad9b2057bfe486157c)

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -951,6 +1269,73 @@ Identity SDFFeatures::ConstructSdfCollision(
+    std::size_t modelID = worldModel->nestedModelNameToEntityId.at(modelName);
+    auto model = this->models.at(modelID);
+
+    model->body->setFixedBase(true);
```
the `btMultiBody::setFixedBase` API was added in https://github.com/bulletphysics/bullet3/commit/19aafd522114906c0f4a35a5fcde58400e382f5d and released in bullet versions 2.89 and 3.05. Ubuntu 20.04 CI is failing as it only has bullet version 2.88 (22.04 CI looks fine). If this API is needed, we may need to retarget to Harmonic

iche033 replied:
added `ifdef` for bullet version in 20.04. `ConstructSdfJoint` will not work in versions < 2.89. [88962e8](https://github.com/gazebosim/gz-physics/pull/574/commits/88962e84cda332c69e27fa5e83e2dfc2d1427d16)

scpeters replied:
I've proposed using 22.04 for GitHub workflow code coverage testing in #578, which should give use us better insight into the coverage of this feature

iche033 replied:
expanded test to cover the construct joint feature. [15aab54](https://github.com/gazebosim/gz-physics/pull/574/commits/15aab54d1e02a0b1e618b346ed19e48726ae6a6c)

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -317,63 +419,214 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
+    // multiple subt-trees detected in model
+    // \todo(iche033) support multiple kinematic trees and
+    // multiple floating links in a single model
+    gzerr << "Multiple subt-trees / floating links detected in a model. "
```
```suggestion
    gzerr << "Multiple sub-trees / floating links detected in a model. "
```

iche033 replied:
[26471eb](https://github.com/gazebosim/gz-physics/pull/574/commits/26471eb4da6a791a741a7716de706d1ac79f07b1)

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -317,63 +419,214 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
-  const auto *world = this->ReferenceInterface<WorldInfo>(_worldID);
+  if (structures.size() > 1)
+  {
+    // multiple subt-trees detected in model
```
```suggestion
    // multiple sub-trees detected in model
```

iche033 replied:
[26471eb](https://github.com/gazebosim/gz-physics/pull/574/commits/26471eb4da6a791a741a7716de706d1ac79f07b1)

mjcarroll commented on bullet-featherstone/src/Base.hh:
```diff
@@ -99,7 +101,7 @@ struct ModelInfo
     Identity _world,
     Eigen::Isometry3d _baseInertiaToLinkFrame,
-    std::unique_ptr<btMultiBody> _body)
+    std::shared_ptr<btMultiBody> _body)
```
Why was this ownership change (here and elsewhere) necessary?  Is a unique_ptr not still sufficient here?

iche033 replied:
yeah, all we now use a single `btMultiBody` to store links from both the top level model and also nested models. So the nested model now keeps a shared pointer to it.

scpeters commented on test/common_test/world_features.cc:
```diff
@@ -16,6 +16,8 @@
 */
 #include <gtest/gtest.h>
 
+#include <LinearMath/btScalar.h>
```
I think we don't want to include a bullet header file in a common test

scpeters replied:
oh, I see it's in order to use the `btGetVersion()` API. I think I would prefer detecting the version in cmake and use `target_compile_definitions` to define a macro indicating whether the version is `>= 2.89` or `< 2.89`

iche033 replied:
removed header and updated other places in test dir that use `btGetVersion()`. It looks like our `FindGzBullet` does not set a `GzBullet_VERSION` variable or `Bullet_VERSION`. So I added extra logic in the test cmake file to get the bullet version using `pkg_check_module`. We can also move that logic to gz-cmake later.

scpeters replied:
nit: my favorite way to do this is to do the version check in cmake with its [version comparisons](https://cmake.org/cmake/help/latest/command/if.html#version-comparisons) (like `VERSION_GREATER_EQUAL`), and then use version comparison in the macro name, like `BT_BULLET_VERSION_GE_289` (see an [example from gz-common](https://github.com/gazebosim/gz-common/blob/983a1eb4bec2753c7c1313534c99d26474a3a9c4/graphics/src/CMakeLists.txt#L79-L85)). This is a little more robust if there was a patch version added for example

iche033 replied:
updated checks, added definitions for `BT_BULLET_VERSION_LE_325` and `BT_BULLET_VERSION_GE_289` [b963af1](https://github.com/gazebosim/gz-physics/pull/574/commits/b963af17b666bcbe5c8ea10a16f6cbe07d8ad651)`

scpeters commented:
I just built this against `gz-sim7` and when running the [nested_model_joint_positions.sdf](https://github.com/gazebosim/gz-sim/blob/gz-sim7/examples/worlds/nested_model_joint_positions.sdf), it seems that bullet-featherstone is not putting `model4` in the correct initial pose.

### dart: nested_model_joint_positions.sdf

~~~
gz sim -v 4 src/gz-sim/examples/worlds/nested_model_joint_positions.sdf
~~~

![Screenshot 2023-11-30 at 4 33 06 PM](https://github.com/gazebosim/gz-physics/assets/3650755/c8672650-b6cb-4e9c-9d0f-0979c52c5107)

`model4` is on the far right

### bullet-featherstone: nested_model_joint_positions.sdf

~~~
gz sim -v 4 src/gz-sim/examples/worlds/nested_model_joint_positions.sdf --physics-engine gz-physics-bullet-featherstone-plugin
~~~

![Screenshot 2023-11-30 at 4 33 31 PM](https://github.com/gazebosim/gz-physics/assets/3650755/a7725902-26ce-476c-ae16-4e30be064421)

`model4` is now next to `model1` on the far left, but sunk into the ground and bounces away when the simulation starts

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -143,167 +146,266 @@ void extractInertial(
+/// \param[in] _model Parent model of joint
+/// \param[in] _joint Joint name
+/// \param[in] _link Scoped link name
+::sdf::Errors resolveJointPoseRelToLink(math::Pose3d &_resolvedPose,
```
I think the logic in this function can be simplified. I've started trying to do so, but it's not working yet

iche033 replied:
ok thanks for the tip. I actually tried it before but couldn't get it to work and now I found that it was because I wasn't passing the fully scoped name.  Code should be simpler now. [575e73b](https://github.com/gazebosim/gz-physics/pull/574/commits/575e73b587c4f4523bdc97b00be7423dd17391c5)

iche033 commented on bullet-featherstone/src/FreeGroupFeatures.cc:
```diff
@@ -33,6 +33,11 @@ Identity FreeGroupFeatures::FindFreeGroupForModel(
     return this->GenerateInvalidId();
 
+  // bullet-featherstone does not allow floating bodies so if a joint exists
+  // the multibody does not quality as a FreeGroup
```
ah I misunderstood the meaning of a freegroup. I removed this check and comment. [575e73b](https://github.com/gazebosim/gz-physics/pull/574/commits/575e73b587c4f4523bdc97b00be7423dd17391c5)

iche033 commented:
> model4 is now next to model1 on the far left, but sunk into the ground and bounces away when the simulation starts

Fixed a todo item and that resolved the pose issue of model4 which occurs when the root link is inside a nested model (as opposed to a top level model). [575e73b](https://github.com/gazebosim/gz-physics/pull/574/commits/575e73b587c4f4523bdc97b00be7423dd17391c5)

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -958,6 +1257,88 @@ Identity SDFFeatures::ConstructSdfCollision(
+  // >= 2.89. This is needed for dynamically creating world joints,
+  // i.e. setting the btMultiBody to be fixed. So output an error letting
+  // users know the joint will not be created.
+  // \todo(iche033) An workaround for this is to loop through all the joints
```
```suggestion
  // \todo(iche033) A workaround for this is to loop through all the joints
```

scpeters commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -958,6 +1257,88 @@ Identity SDFFeatures::ConstructSdfCollision(
+    childLinkName = _sdfJoint.ChildName();
+  }
+
+  // Currently only supports constructing joint with world
```
```suggestion
  // Currently only supports constructing fixed joint with world as parent
```

iche033 replied:
[da3aac6](https://github.com/gazebosim/gz-physics/pull/574/commits/da3aac6ffbaa02faf2b77635caa21bd868aa505e)

scpeters reviewed (commented):
overall it looks good; I left a few minor comments and will approve once they are resolved

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -586,21 +794,63 @@ Identity SDFFeatures::ConstructSdfModel(
     return this->GenerateInvalidId();
 
+  auto modelToNestedModel = Eigen::Isometry3d::Identity();
+  // if the model of the root link is in a top level model, compute
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -143,167 +146,207 @@ void extractInertial(
-          rootLink = child;
-          rootJoint = joint;
-          fixed = true;
+  // compute joint pose relative to parent link
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -461,53 +676,46 @@ Identity SDFFeatures::ConstructSdfModel(
+      btTransform offsetInABt, offsetInBBt;
+      offsetInABt = parentLocalInertialFrame * parent2jointBt;
+      offsetInBBt =
+         convertTf(linkToComTf.inverse() * poseJointToChild.inverse());
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -317,63 +360,214 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
+  {
+    // multiple sub-trees detected in model
+    // \todo(iche033) support multiple kinematic trees and
+    // multiple floating links in a single model
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -143,167 +146,207 @@ void extractInertial(
+/// \brief Find all the root links given a model SDF
+/// \param[in] _sdfModel Model SDF
+/// \param[in] _parentOf A map of child link to its parent info
+/// \param[out] _rootLinks A vector of root links and its immediate parent model
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -317,63 +360,214 @@ std::optional<Structure> ValidateModel(const ::sdf::Model &_sdfModel)
+          << std::endl;
+  }
+  // Create model for the first structure.
+  auto structure = structures[0];
```
[ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -461,53 +676,46 @@ Identity SDFFeatures::ConstructSdfModel(
+      btTransform parentLocalInertialFrame = convertTf(
+        parentLinkInfo->inertiaToLinkFrame);
+      btTransform parent2jointBt = convertTf(
+        // gz::math::eigen3::convert(parent2joint.Inverse()));  // X_PJ
```
removed. [ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/Base.hh:
```diff
@@ -357,6 +387,99 @@ class Base : public Implements3d<FeatureList<Feature>>
+      }
+    }
+
+    // If nested, we are done here. No need to remove multibody
```
ah yes, I added a todo comment. [ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented on bullet-featherstone/src/EntityManagementFeatures.cc:
```diff
@@ -206,53 +206,16 @@ Identity EntityManagementFeatures::ConstructEmptyWorld(
-  auto *world = this->ReferenceInterface<WorldInfo>(model->world);
-  if (world->modelIndexToEntityId.erase(model->indexInWorld) == 0)
-  {
-    // The model has already been removed at some point.
+  if (!model)
     return false;
```
removed. [ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

scpeters replied:
well, now I think I was wrong, since I didn't notice that we are dereferencing `model` on the next line. Those two lines from https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339 should probably come back. Thanks for your patience

iche033 replied:
ah yes, reverted change. [d08d0ef](https://github.com/gazebosim/gz-physics/pull/574/commits/d08d0ef83a4e4cc638ff11eb792877be90a0ea9f)

iche033 commented on bullet-featherstone/src/SDFFeatures.cc:
```diff
@@ -143,167 +146,207 @@ void extractInertial(
 
-  for (std::size_t i = 0; i < _sdfModel.ModelCount(); ++i)
+/////////////////////////////////////////////////
+/// \brief Recursively build a tree of parent-child data structures from the
```
updated comment. [ca1715e](https://github.com/gazebosim/gz-physics/pull/574/commits/ca1715e2a41ecb070fe8d81d7530f75e0c985339)

iche033 commented:
> overall it looks good; I left a few minor comments and will approve once they are resolved

thanks for the review! Addressed all comments.

scpeters commented:
also, one small patch to improve test coverage:

~~~
diff --git a/test/common_test/world_features.cc b/test/common_test/world_features.cc
index 640e185c..e1e68c0f 100644
--- a/test/common_test/world_features.cc
+++ b/test/common_test/world_features.cc
@@ -434,6 +434,10 @@ TEST_F(WorldNestedModelTest, WorldConstructNestedModel)
     gz::physics::ForwardStep::State state;
     gz::physics::ForwardStep::Output output;
 
+    // Check invalid input to RemoveNestedModel method
+    EXPECT_FALSE(worldModel->RemoveNestedModel(1));
+    EXPECT_FALSE(worldModel->RemoveNestedModel("invalid"));
+
     // Check that we can remove models via RemoveNestedModel
     EXPECT_TRUE(worldModel->RemoveNestedModel(0));
     EXPECT_TRUE(nestedModel->Removed());
~~~

scpeters reviewed (approved):
just two remaining comments: a suggested revert of one of my suggestions and a patch to improve coverage

iche033 commented:
> also, one small patch to improve test coverage:

Added, thanks. [d08d0ef](https://github.com/gazebosim/gz-physics/pull/574/commits/d08d0ef83a4e4cc638ff11eb792877be90a0ea9f)
